### ISSUE TITLE
Parallel Index and Stat Maintenance Jobs

### ISSUE THREAD

**Issue by fredclown** [olahallengren/sql-server-maintenance-solution#274]
It would be nice to have an option to have the Index and Stat maintenance run parallel operations. Currently, it is all done serially. I could envision maybe a parameter in the setup script to create multiple worker jobs and a controller job. The controller job would create a queue of sorts and the then kick off the workers that would then deplete the queue and stop when the queue has been completed. For large databases this would be a huge win for performance and the ability to just get through everything prior to the next schedule kicking off.

**Comment by EitanBlumin** [comment]
This should be possible to implement by taking the `@tmpIndexesStatistics` table and turning it into a "Queue-like" table (similar to the `QueueDatabase` table), instead of repopulating it on every execution.

Then, it should be possible to add a parameter which, in tandem with `@DatabasesInParallel` will determine the "level" of parallelism (i.e. at the database level, schema level, object level, index level, or partition level).

It will definitely be extremely useful for databases with a huge number of objects.

**Comment by fredclown** [comment]
Yeah, trying to use the index maintenance on a HUGE database serially is just not possible. For instance we have a database that is about 2TB in size. I took over 24 hours to do the index maintenance one at a time, but the server had CPU to spare. So, it could handle more than one thread. So, I've had to create multiple index maintenance jobs and try and partition out the indexes evenly among the jobs via the page sizes. Lot's of trial an error to try and get the timings to be about the same per job. Even so there are days where one goes longer than the others and its a constant tuning maintenance task.

**Comment by EitanBlumin** [comment]
Thinking about it, it would also be very beneficial to make the `DatabaseIntegrityCheck` procedure have a similar capability. Assuming the `@CheckCommands` parameter receives CHECKTABLE or CHECKFILEGROUP values, it should be possible to control its "level" of parallelism as well (i.e. at the database level, schema level, table level, or file-group level).

I'll create a separate issue for that.

**Comment by edo1** [comment]
> Yeah, trying to use the index maintenance on a HUGE database serially is just not possible

Agreed, especially with the standard edition where the MAXDOP option is not honored.

@olahallengren is there any chance to get this feature integrated?

**Comment by fredclown** [comment]
Since it wasn't in Ola's script (which is an excellent script) I built one that does parallel index defragmentation. It uses many of the same parameters as Ola's script.

[https://bitbucket.org/davinstuder/index-maintenance/src/master/](https://bitbucket.org/davinstuder/index-maintenance/src/master/)
